### ISSUE TITLE
Refactor capstone2llvmir

### ISSUE THREAD

**Issue by PeterMatula** [avast-tl/retdec#115]
The current implementation is an early prototype. We should refactor it in order to make it prettier and easier to use and develop for others. Do the following:
- Separate public API that is needed for users from implementation API that is needed for developers.
- Refactor code.
  - Solve todos.
  - Move common code to parent classes.
  - Add public API function to allow instruction-by-instruction decoding.
  - Throw different kinds of exceptions.
- Unit tests for x86 and x86_64 registers according to issues discussed [here](https://stackoverflow.com/questions/25455447/x86-64-registers-rax-eax-ax-al-overwriting-full-register-contents?noredirect=1&lq=1).
- Add auto generation of unhandled instructions based on information from Capstone -> generate pseudo calls of functions named as ASM instructions. 
- Add configuration API - enable/disable throwing asserts, exceptions, unknown instructions handling, etc.

**Comment by PeterMatula** [comment]
Make sure `rep movsb` (and other such instructions) are ok. This is an older report, it might be fine now. If so, at least add unit or regression test.

This ASM code:
```asm
0x403975:   be d8 42 40 00                      mov esi, 0x4042d8
0x40397a:   83 c6 05                            add esi, 0x5
0x40397d:   b9 04 00 00 00                      mov ecx, 0x4
0x403982:   bf fe 42 40 00                      mov edi, 0x4042fe
0x403987:   f3 a4                               rep movsb
0x403989:   46                                  inc esi
0x40398a:   b9 04 00 00 00                      mov ecx, 0x4
0x40398f:   f3 a4                               rep movsb
0x403991:   46                                  inc esi
0x403992:   b9 04 00 00 00                      mov ecx, 0x4
0x403997:   f3 a4                               rep movsb
0x403999:   46                                  inc esi
0x40399a:   b9 04 00 00 00                      mov ecx, 0x4
0x40399f:   f3 a4                               rep movsb
```

Gets decompiled to (at the time of report):
```c
char * g110; // 0x4042d8
char * g111; // 0x4042dd
char * g112; // 0x4042de
char * g113; // 0x4042df
char * g114; // 0x4042e0
...
char * g117; // 0x4042fe
...
memcpy((char *)&g117, (char *)&g111, 4);
memcpy((char *)&g117, (char *)&g112, 4);
memcpy((char *)&g117, (char *)&g113, 4);
g4 = (int32_t)memcpy((char *)&g117, (char *)&g114, 4);
```
Where `edi` and `esi` are set to original values after `rep movsb` - which does not happen on CPU.

It should be something like:
```c
memcpy(g117, g111, 4);
memcpy(g117+4, g111+5, 4);
memcpy(g117+8, g111+10, 4);
memcpy(g117+12, g111+15, 4);
```
Where second operands get increased by 5 bytes, not one, and first operands by 4 bytes, not 0.

**Comment by PeterMatula** [comment]
TODO for me: write a few words about the basic principles to wiki once this is done. Something like [this comment](https://github.com/avast-tl/retdec/issues/193#issuecomment-363799544) in #193.

**Comment by PeterMatula** [comment]
FIxed by c2ed626a7027a17fe6c6f0781a162751efeec6bb.
There probably will be further improvements, but the most important things were addresses.
